---
title: "**RNA-Seq Data Visualization & Analysis with R**"
author: "Matt Maxwell"
date: "`r format(Sys.Date(), '%B %d, %Y')`"
output: 
  html_document: 
   code folding: hide
   hard_wrap: TRUE
   theme: lumen
   toc: TRUE 
   toc_float: TRUE
   fig_caption: yes

---


<style>
.tocify {
  font-size: 11pt;
}

p.caption {
  font-size: 1.25em;
  color: black
}

pre {
  overflow-x: scroll
}

pre code {
  white-space: pre;
}

body {
  font-size: 12pt;
}

</style>

### **RNA-seq analysis pipeline**
The RNA-seq analysis pipeline sections colored in gray will not be covered here but were performed using bash shell scripting and the [HOMER bioinformatics suite](http://homer.ucsd.edu/homer/) on the Salk Institute Integrative Genomics & Bioinformatics computing server. The pipeline sections highlighted in pink are performed in Rstudio and are covered in depth here including the R code and resulting outputs.
  RNA-seq analyses such as these have been critical tools in my doctoral work looking to elucidate the molecular consequences of loss of function ARID1A tumor suppressor gene mutations. For example, through GSEA analyses I found that Interferon Alpha Response (Type I Interferon) gene set signature activation is a common consequence of ARID1A mutation across multiple different cancer cell types which formed the foundation of my thesis work.


```{r RNA-seq-Diagram, echo=FALSE}
library(DiagrammeR)
library(htmlwidgets)
flowchart <- grViz("
      digraph flowchart {
      # node definitions with substituted label text
      node [fontname = Helvetica, shape = rectangle, style = filled, fillcolor = gray90]        
      tab1 [label = '@@1', fillcolor=gray90]
      tab2 [label = '@@2', fillcolor=gray90]
      tab3 [label = '@@3', fillcolor=gray90]
      tab4 [label = '@@4', fillcolor=plum1]
      tab5 [label = '@@5', fillcolor=plum1]
      tab6 [label = '@@6', fillcolor=plum1]
      tab7 [label = '@@7', fillcolor=plum1]
      tab8 [label = '@@8', fillcolor=plum1]




      # edge definitions with the node IDs
      tab1 -> tab2 -> tab3 -> tab4;
      tab4 -> tab5;
      tab4 -> tab6;
      tab4 -> tab7;
      tab4 -> tab8;

      }

      [1]: 'Align sequencing files to mouse (mm10) reference genome using STAR'
      [2]: 'Generate a raw transcript counts table using DESEQ2'
      [3]: 'Perform differential expression analysis using edgeR'
      [4]: 'Data visualization'
      [5]: 'Volcano plot'
      [6]: 'Heatmap of differentially expressed genes'
      [7]: 'GSEA Dotplot'
      [8]: 'Venn Diagram'

      ")
flowchart
```


## **Import data & load packages**
The data visualized & analyzed here is from an RNA-seq experiment comparing the transcriptomes of MC38 mouse colon cancer cells genetically engineered with CRISPR-Cas9 to be deficient in the ARID1A tumor suppressor gene (sg*Arid1a*) relative to CRISPR control MC38 cells that are *Arid1a* wildtype (sgScramble). 
  The `MC38_edgeR_df` dataframe from this experiment includes important data such as normalized mRNA transcript counts, Log2fc, and adjusted p values across the transcriptome.
```{r import-data-load-packages, message=FALSE, hard_wrap=TRUE, warning=FALSE}
#To set working directory on windows machine, replace backslashes with double backslashes for R to accept our file path
wd <- r"(C:\Users\mattm\OneDrive\Desktop\GitHub_projects\RNA_Seq_Data_Visualization_with_R)"

#set working directory
setwd(wd)

#set this as our working directory for all R code chunks in this tutorial.
#IGNORE if you're not using R markdown file
knitr::opts_chunk$set(root.dir = wd)

#Set preference for no warnings to be given when running code chunks in R markdown
#IGNORE if you're not using R markdown file
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 

#read in our dataframe
MC38_edgeR_df <- read.delim("sgArid1a_MC38_edgeR.txt", header = TRUE, sep = "\t")

#load packages with package manager package called pacman
if (!require("pacman")) install.packages("pacman")
pacman::p_load(here,  
               tidyverse, #dyplyr and ggplot utilities
               scales, # Transform axis scales   
               ggrepel, # Optimize plot label separation 
               ComplexHeatmap, #Awesome heatmap package
               viridis, #Package for some of my favorite color palettes like Magma
               RColorBrewer,#Package for some of my favorite color palettes like RdBu
               forcats, #Package for working with factors
               msigdbr, #Molecular signatures database package for GSEA
               clusterProfiler, #Package for GSEA dotplot 
               euler) #Package for making venn diagrams
```


## **Data wrangling for volcano plot**
To make a volcano plot, we need three pieces of data: 1) gene name, 2) Log2 fold change, and 3) adjusted p values. Here, I demonstrate how to grab these three pieces of data and use them make a volcano plot that highlights genes whose expression significantly changes as well as to specifically highlight some individual genes I think are interesting.
```{r, Data-wrangling-volcano-plot, message=FALSE}
#Rename our columns of interest using the rename function in dyplyr from tidyverse package
MC38_edgeR_df <- rename(MC38_edgeR_df, Gene = Annotation.Divergence, Log2fc = sgArid1a.vs..sgScramble.Log2.Fold.Change, adj_p_value =sgArid1a.vs..sgScramble.adj..p.value)


#Grab the gene name values preceding "|" edgeR from Gene column 
MC38_edgeR_df$Gene <- sub("\\|.*", "", MC38_edgeR_df$Gene)

#grab the columns we need for volcano plot
volcano_df <- MC38_edgeR_df[, c("Gene", "Log2fc", "adj_p_value")]


# Replace NA values in Log2FC column with Zero.
volcano_df$Log2fc[is.na(volcano_df$Log2fc)] <- 0


#Define the parameters you want to classify a gene as upregulated, donwregulated, or not significantly changed (NS)
volcano_df <- volcano_df %>%
  mutate(gene_type = case_when(Log2fc >= .585 & adj_p_value <= 0.05 ~ "Upregulated",
                               Log2fc <= -.585 & adj_p_value <= 0.05 ~ "Downregulated",
                               TRUE ~ "NS"))

#Count the number of genes in the three classes we defined
volcano_df %>%
  count(gene_type)

#Specify colors, sizes, and transparancy values associated with the three classes 
cols <- c("Upregulated" = "#ffad73", "Downregulated" = "#26b3ff", "NS" = "grey") 
sizes <- c("Upregulated" = 3, "Downregulated" = 3, "NS" = 1) 
alphas <- c("Upregulated" = 1, "Downregulated" = 1, "NS" = 0.5)

# Define genes to highlight on the volcano plot
signif_genes <- volcano_df %>%
  filter(Gene %in% c("Tap1", "Cxcl10", "Stat1", "Ptgs1", "Cd47", "Il4ra"))

Up_genes <- volcano_df %>%
  filter(Gene %in% c("Tap1", "Cxcl10", "Stat1"))

Down_genes <- volcano_df %>%
  filter(Gene %in% c("Ptgs1", "Cd47"))

```


## **Volcano plot**
Here, we use [ggplot from the Tidyverse R package](https://ggplot2.tidyverse.org/) to make the volcano plot because ggplot allows more room for customization when making a volcano plot compared to some other R packages.
```{r volcano-plot, fig.cap = "**Figure 1: Volcano plot Highlighting Differentially Expressed Genes in sgArid1a MC38 cells**", message=FALSE}

#Customize what we want to show in our the volcano plot
final_plot <- ggplot(data = volcano_df,
                     aes(x = Log2fc,
                         y = -log10(adj_p_value))) + 
  geom_point(aes(colour = gene_type), 
             alpha = .75, 
             shape = 16,
             size = 4) + 
  geom_point(data = Up_genes,
             shape = 21,
             size = 4, 
             fill = "firebrick", 
             colour = "black") + 
  geom_point(data = Down_genes,
             shape = 21,
             size = 4, 
             fill = "steelblue", 
             colour = "black") + 
  geom_hline(yintercept = -log10(0.05),
             linetype = "dashed") +
  geom_vline(xintercept = c(log2(.75), log2(1.5)),
             linetype = "dashed") +
  geom_label_repel(data = signif_genes,   
                   aes(label = Gene),
                   force = 1,
                   nudge_y = 2) +
  scale_colour_manual(values = cols) + 
  scale_x_continuous(breaks = c(seq(-6, 7, 2)),     
                     limits = c(-6, 7)) +
  labs(title = "MC38 Volcano Plot", 
       x = "Log2FC (sgArid1a/sgScramble)",
       y = "-log10(adjusted P-value)",
       colour = "Gene type") +
  theme_bw() + # Select theme with a white background  
  theme(panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5),    
        panel.grid.minor = element_blank(),
        panel.grid.major = element_blank())

#Call the volcano plot
final_plot
```
Volcano plots such as this are a great way to visualize pair wise comparisons of the number of genes changing, the magnitude of change, and statistical confidence associated with those changes. It can also be useful when presenting data to highlight genes of interest on our volcano plot such as the three up & downregulated genes I've highlighted on this volcano plot. For context, six genes I've highlighted all relate to the process of anti-tumor immunity which is the process I'm studying and thus of value to highlight on the volcano plot to communicate these specific changes.

## **Data wrangling for Heatmap**
To make a heatmap, we need to 1) grab the normalized counts values for each sample from our edgeR dataframe (MC38_edgeR_df), 2) filter for genes we want to visualize such as genes whose expression significantly changes, and 3) convert this subset of genes from a dataframe to a matrix to use as input for the ComplexHeatmap. In addition, we will also utilize some code to highlight genes of interest (same genes of interest on volcano plot) on the heatmaps we make.
```{r data-wrangling-heatmap, message=FALSE}
#Rename the normalized counts columns we'll need for our heatmap by referencing their column number
MC38_edgeR_df <- rename(MC38_edgeR_df, sgScr_1 = 9, sgScr_2 = 10, sgArid1a_1 = 11, sgArid1a_2 = 12)

#Filter for significantly up or downregulated genes
#We only want to plot significantly changed genes on this heatmap
MC38_edgeR_df2 <- MC38_edgeR_df %>%
  filter(Log2fc >= .585 |Log2fc <= -.585, adj_p_value <= .05)

#Grab gene name and edgeR normalized counts for each group
heatmap_df <- MC38_edgeR_df2[, c(8:12)]

#Make dataframe without gene column, the first column of the df
heat_mat <- heatmap_df[,-1]

#Assign gene names from 'heatdata' as row names to new dataframe object 'mat'
rownames(heat_mat) <- heatmap_df[,1]

#Convert dataframe to a matrix, can use data.matrix() or as.matrix()
heat_mat <- data.matrix(heat_mat)

#Generate gene Z scores and transpose matrix
heat_mat <- t(scale(t(heat_mat))) 

#List of interesting genes we'd like to label on heatmap
heat_anotation <- signif_genes

#add a column for row # in heatmap_df
heatmap_df$rownumber = 1:nrow(heatmap_df)

#Add row number values from heatdata to heat_anotation
heat_anotation <- merge(heat_anotation,heatmap_df, all = F)

#Print row numbers for heat_anotation from heatdata
cat(heat_anotation$rownumber,sep=",")


x <- heat_anotation$Gene
x_list <- as.list(x, ",")
print(x_list)

#Create object with row name locations of genes from heatdata that you'd like to annotate on heatmap
genelabels = rowAnnotation(foo = anno_mark(at = c(229,234,215,352,295,16),
                                           labels = x_list))




```


## **ComplexHeatmap - RdBu**
To make the heatmap from our data matrix, we simply need to use the Heatmap function from the [ComplexHeatmap package](https://jokergoo.github.io/ComplexHeatmap-reference/book/index.html). However, I also want to make some additional asethetic modifications to my heatmap such as giving each genotype their own color coded top label and specifying that the heatmap utilize the "RdBu" color scheme from the RColorBrewer R package.
```{r heatmap, fig.cap = "**Figure 2: Heatmap of Differentially Expressed Genes in sgArid1a MC38 cells**", fig.height=9, message=FALSE}

#Define a list of our sample genotypes as a factor of either sgScramble or sgArid1a
fa = factor(c("sgScramble", "sgScramble", "sgArid1a", "sgArid1a"),
            levels = c("sgScramble", "sgArid1a"))

#Define Heatmap annotation color bars for genotype designation 
ha = HeatmapAnnotation(Genotype = fa, height = unit(1, "cm"),
                       col = list("Genotype"=c("sgScramble"="black","sgArid1a"="orange2")))

#Passing my fav color palette RdBu to mypalette object
mypalette = rev((brewer.pal(n=9, ("RdBu"))))

ht <- Heatmap(heat_mat, 
              col = mypalette, 
              top_annotation = ha, heatmap_width = unit(6 , "cm"), heatmap_height = unit(16, "cm"),
              border = TRUE, cluster_rows = T, right_annotation = genelabels,
              show_column_dend = F, show_row_dend = F, show_row_names = F,  column_names_side = "top", 
              column_names_rot = 2, show_column_names = F, cluster_columns = F,
              use_raster=T, row_km = 1, raster_quality=5, 
              column_names_gp = gpar(fontsize=10),
              heatmap_legend_param = list(height = unit(16, "cm"), direction = "horizontal", 
                                          title = "mRNA Expression Z Score", border = "black", 
                                          title_position = "topcenter")) 
ht_legend_dn <- draw(ht, heatmap_legend_side = "bottom")



```
Heatmaps are useful for visualizing the relative relationships between gene expression values between samples where rows are individual genes and columns are samples. You can also communicate how many genes are upregulated versus downregulated in relative terms with heatmaps. Here, I've highlighted the same six genes I highlighted in the volcano plot that I know are important for the biological process I'm studying, anti-tumor immunity.

## **ComplexHeatmap - Magma**
If you want a color palette that's not offered in the RColorBrewer package, you could also use the Viridis color palette package which has awesome palettes like "Magma". Here's the same heatmap with the Magma color palette.  
```{r Magma-Heatmap, fig.cap = "**Figure 3: Heatmap with Magma color palette from Viridis package**", fig.height=9, message=FALSE}

ht <- Heatmap(heat_mat, 
              col = viridisLite::magma(n = 100), 
              #col = mypalette, 
              top_annotation = ha, heatmap_width = unit(6 , "cm"), heatmap_height = unit(16, "cm"),
              border = TRUE, cluster_rows = T, right_annotation = genelabels,
              show_column_dend = F, show_row_dend = F, show_row_names = F,  column_names_side = "top", 
              column_names_rot = 2, show_column_names = F, cluster_columns = F,
              use_raster=T, row_km = 1, raster_quality=5, 
              column_names_gp = gpar(fontsize=10),
              heatmap_legend_param = list(height = unit(16, "cm"), direction = "horizontal", 
                                          title = "mRNA Expression Z Score", border = "black", 
                                          title_position = "topcenter")) 
ht_legend_dn <- draw(ht, heatmap_legend_side = "bottom")
```


## **Data wrangling for GSEA Dotplot**
To provide an informative conceptualization of which biological processes are changed in our experiment, we'll use gene set enrichment analysis (GSEA) to identify which biological processes are enriched in our up or downregulated genes in ARID1A deficient MC38 cells. Similar to the volcano plot, we need to grab gene name, Log2FC, and adjusted p value but only of genes whose expression is significantly changed. We then perform GSEA analysis using the [ClusterProfiler R package](https://yulab-smu.top/biomedical-knowledge-mining-book/index.html) and plot the results using ggplot.
```{r, data-wrangling-gsea-dotplot, message=FALSE}
#Grab gene name, Log2fc, and adj_p_value
GSEA_df <- volcano_df[,1:3]

#Add a column for GSEA values
GSEA_df$GSEA <- -log10(GSEA_df$adj_p_value) * sign(GSEA_df$Log2fc)

#Create object with gene name and GSEA value
geneList <- GSEA_df[,4]
names(geneList) <- as.character(GSEA_df[,1])

#sort the genes in the list according to highest to lowest GSEA value
geneList <- sort(geneList,decreasing=TRUE)

#Lets see the top genes in the list
head(geneList)

#Use msigdbr package to read in the gene sets for GSEA analysis
#msigdbr gives access to many different gene sets
#For example, instead of Hallmarks, could import GO gene sets by specificying category ="C5"
Hallmark <- msigdbr(species = "Mus musculus", category = "H")
Hallmark_v2 = Hallmark %>% dplyr::select(gs_name, gene_symbol) %>% as.data.frame()

#Run GSEA analysis of your gene list against Hallmarks
set.seed(8888)
Gsea_result <- GSEA(geneList, TERM2GENE=Hallmark_v2, verbose=FALSE,
                    pvalueCutoff = .8, pAdjustMethod = "BH", nPerm = 1000)

# Plot only selected gene sets
library(stringr)

## count the gene number for each enrichment
Gene_count<- Gsea_result@result %>% group_by(ID) %>% 
  summarise(count = sum(str_count(core_enrichment,"/")) + 1)

## merge with the original dataframe
GSEA_data_frame <- left_join(Gsea_result@result, Gene_count, by = "ID") %>% mutate(GeneRatio = count/setSize)

## for reordering the factor
library(forcats) 
Data_activated1 <- GSEA_data_frame %>% filter(NES>0 | NES < 0)

#Take the word Hallmark out of gene sets for plot visualization purposes
Data_activated1$Description <- gsub("_", " ", Data_activated1$Description)
Data_activated1$Description <- str_to_title(Data_activated1$Description)
Data_activated1$Description <- gsub("Hallmark ", "", Data_activated1$Description)


#Arrange gene set df by NES score
Data_activated1 <- arrange(Data_activated1, desc(NES))

#Take the hightest and lowest NES value gene sets 
upregulated_gs <- slice_head(Data_activated1, n=5)
downregulated_gs <- slice_tail(Data_activated1, n=5)

#Combine the upregulated_gs and downregulalted_gs objects to plot on dotplot
Dotplot_gs <- rbind(upregulated_gs, downregulated_gs)





```


## **GSEA Dotplot**
Plot the GSEA dotplot using ggplot.
```{r GSEA-dotplot, fig.cap = "**Figure 4: GSEA Dotplot of sgArid1a Upregulated Genes in MC38 cells**", message=FALSE}

# ggplot(Data_activated1, aes(NES, fct_reorder(Description, NES), fill=Data_activated1$p.adjust), showCategory=5) +
#     geom_col(orientation='y') +
#     scale_fill_continuous(low='red', high='blue', guide=guide_colorbar(reverse=TRUE)) +
#     theme_minimal() + ylab(NULL)

custom_dotplot <- ggplot(Dotplot_gs, aes(NES, fct_reorder(Description, NES))) +
  geom_point(aes(size = GeneRatio, color = p.adjust)) +
  theme_minimal(base_size = 15) +
  #scale_colour_viridis_c(limits = c(5.0e-06, 2.5e-05). direction = 1, option = "viridis")+
  ylab(NULL) +
  #ggtitle("Hallmark enrichment", ) +
  scale_size_continuous(range = c(3, 14)) + scale_colour_gradient(low="red", high = "blue") + coord_cartesian(clip = "off") +
  labs(color="Adjusted p-value", size="Gene Set Ratio") +
  theme(axis.text=element_text(size=10,color="black")) +
  xlab("NES (sgArid1a/sgScramble)") +
  ggtitle(substitute(paste(bold("GSEA Hallmarks Dotplot")))) +
  theme(axis.title.x = element_text(size = 15, angle = 0, vjust = -1))
custom_dotplot
```


## **Venn Diagram**
We can also make venn diagrams to assess the degree of overlap between two or more sets of genes of interest using the R package [euler](https://cran.r-project.org/web/packages/eulerr/vignettes/introduction.html) which generates proportionally accurate sized venn diagrams unlike other common web-based venn diagram makers such as [venny](https://bioinfogp.cnb.csic.es/tools/venny/). For example, I'm curious to know the degree of overlap between ARID1A deficient MC38 colon cancer cells and ARID1A deficient B16F10 melanoma cell and will make a venn diagram below to get my answer.
```{r}
#Filter for MC38 upregulated genes
MC38_upregulated <- MC38_edgeR_df %>%
  filter(Log2fc >= .585, adj_p_value <= .05)

#Import in sgArid1a B16F10 RNA-seq data
B16F10_edgeR_df <- read.delim("sgArid1a_B16F10_edgeR.txt", header = TRUE, sep = "\t")

#replace single periods or double periods in column names with a single underscore
colnames(B16F10_edgeR_df) <- gsub("\\.|\\.\\.", "_", colnames(B16F10_edgeR_df))

#Rename column names in B16F10 edgeR dataframe
B16F10_edgeR_df <- rename(B16F10_edgeR_df, Gene = Annotation_Divergence, Log2fc = sgARID1A_vs_sgNT_Log2_Fold_Change, adj_p_value = sgARID1A_vs_sgNT_adj_p_value)

#Grab the gene name values preceding "|" edgeR from Gene column
B16F10_edgeR_df$Gene <- sub("\\|.*", "", B16F10_edgeR_df$Gene)

#Filter for upregulated B16 genes
B16_upregulated <- B16F10_edgeR_df %>%
  filter(Log2fc >= .585, adj_p_value <= .05)

#Grab gene names in upregulated gene lists
MC38_upregulated <- MC38_upregulated[, "Gene"]
B16_upregulated <- B16_upregulated[, "Gene"]


#Find common and unique genes
common <- intersect(MC38_upregulated, B16_upregulated)
MC38_unique <- setdiff(MC38_upregulated, B16_upregulated)
B16_unique <- setdiff(B16_upregulated, MC38_upregulated)

# Count the number of common entries
num_common <- length(common)

#Euler ven diagram
library("eulerr")
fit <- euler(c("sgArid1a_MC38_up" = 184, "sgArid1a_B16F10_up" = 187,
               "sgArid1a_MC38_up&sgArid1a_B16F10_up" = 65),
             shape = "ellipse")

plot(fit, fills = c("dodgerblue4", "darkgoldenrod1"), font = 9, quantities = TRUE)

```